pkg/apis/autoscaling/types.go

```diff
@@ -95,6 +95,90 @@ type HorizontalPodAutoscalerSpec struct {
 	// more information about how each type of metric must respond.
 	// +optional
 	Metrics []MetricSpec
+
+	// behavior configures the scaling behavior of the target
+	// in both Up and Down directions (scaleUp and scaleDown fields respectively).
+	// If not set, the default HPAScalingRules for scale up and scale down are used.
+	// +optional
+	Behavior *HorizontalPodAutoscalerBehavior
+}
+
+// HorizontalPodAutoscalerBehavior configures a scaling behavior for Up and Down direction
+// (scaleUp and scaleDown fields respectively).
+type HorizontalPodAutoscalerBehavior struct {
+	// scaleUp is scaling policy for scaling Up.
+	// If not set, the default value is the higher of:
```

Comment:
OK.  My only real remaining concern is here.

You are documenting the default behavior as part of the API, but not crystallizing that into the resource itself.

What we normally do is add code to defaults.go, looking for unset values and setting them to the defaults.  This means that a user who did not specify `scaleUp` when they created the object can read the object back and see what the active values are.  It also means that you can have API v3 with a different default if you need to.

This PR is not doing this.  Instead you are assuming defaults deep inside the controller.  If you go this route you can never ever change them because the stored objects have no value set, so any change would be breaking.

I strongly urge you to do it the normal way.  Do you have good reasons to do it differently?

Reply:
I have implemented it as a part of defaults.go
Could you have another look?